pkg/reporter/github_reporter.go

```diff
@@ -0,0 +1,322 @@
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package reporter
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"regexp"
+	"strings"
+	"time"
+
+	"github.com/sethvargo/go-githubactions"
+
+	"github.com/abcxyz/guardian/pkg/flags"
+	"github.com/abcxyz/guardian/pkg/github"
+	"github.com/abcxyz/pkg/logging"
+)
+
+var _ Reporter = (*GitHubReporter)(nil)
+
+const (
+	githubCommentPrefix        = "#### 🔱 Guardian 🔱"
+	githubMaxCommentLength     = 65536
+	githubDestroyIndicatorText = "💥 DESTROY"
+	githubTruncatedMessage     = "\n\n> Message has been truncated. See workflow logs to view the full message."
+)
+
+var githubOperationText = map[Operation]string{
+	PlanOperation:    "PLAN",
+	ApplyOperation:   "APPLY",
+	UnknownOperation: "UNKNOWN",
+}
+
+var githubStatusText = map[Status]string{
+	StartStatus:     "🟨 STARTED",
+	SuccessStatus:   "🟩 SUCCESS",
+	NoChangesStatus: "🟦 NO CHANGES",
+	FailureStatus:   "🟥 FAILED",
+	UnknownStatus:   "⛔️ UNKNOWN",
+}
+
+var (
+	tildeChanged = regexp.MustCompile(
+		"(?m)" + // enable multi-line mode
+			"^([\t ]*)" + // only match tilde at start of line, can lead with tabs or spaces
+			"([~])") // tilde represents changes and needs switched to exclamation for git diff
+
+	swapLeadingWhitespace = regexp.MustCompile(
+		"(?m)" + // enable multi-line mode
+			"^([\t ]*)" + // only match tilde at start of line, can lead with tabs or spaces
+			`((\-(\/\+)*)|(\+(\/\-)*)|(!))`) // match characters to swap whitespace for git diff (+, +/-, -, -/+, !)
+)
+
+// GitHubReporter implements the reporter interface for writing GitHub comments.
+type GitHubReporter struct {
+	gitHubClient github.GitHub
+	flags        flags.GitHubFlags
+	commentID    int64
+	logURL       string
+}
+
+// NewGitHubReporter creates a new GitHubReporter.
+func NewGitHubReporter(ctx context.Context, f flags.GitHubFlags) (Reporter, error) {
+	logger := logging.FromContext(ctx)
+
+	ghCtx, err := githubactions.New().Context()
+	if err != nil {
+		return nil, fmt.Errorf("failed to create github context: %w", err)
+	}
+
+	f.FromGitHubContext(ghCtx)
+
+	if err := validateFlags(f); err != nil {
+		return nil, fmt.Errorf("failed to validate github flags: %w", err)
+	}
+
+	tokenSource, err := f.TokenSource(ctx, map[string]string{
+		"contents":      "read",
+		"pull_requests": "write",
+	})
+	if err != nil {
+		return nil, fmt.Errorf("failed to get token source: %w", err)
+	}
+	token, err := tokenSource.GitHubToken(ctx)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get token: %w", err)
+	}
+
+	gc := github.NewClient(
+		ctx,
+		token,
+		github.WithRetryInitialDelay(1*time.Second),
+		github.WithRetryMaxAttempts(3),
+		github.WithRetryMaxDelay(30*time.Second),
+	)
```

Comment:
I think we want to create the GitHub client with the TokenSource, not the resolved token. The resolved token will expire.

Reply:
So we just mint the token from the source inside the client instead of outside? https://github.com/abcxyz/guardian/blob/main/pkg/github/github.go#L179-L181

Do we have a token source that is capable of generating a new token if it is expired?

Reply:
abcxyz/pkg/githubauth is capable of tokensources that handle refreshes.

Reply:
Spoke offline, this requires a refactor.